### PR TITLE
[5.1] Support for symfony UploadedFile, testing = true

### PR DESCRIPTION
**Pull request by jildertmiedema**
The UploadedFile has a property for testing, this is usefull for testing file upload validation

### CODE REVIEW

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 1]
```diff
@@ -137,6 +145,8 @@ protected function makeRequest($method, $uri, $parameters = [], $cookies = [], $
     {
         $uri = $this->prepareUrlForRequest($uri);
 
+
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: no

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 2]
```diff
@@ -712,4 +730,27 @@ public function dump()
 
         dd($content);
     }
+
+    /**
+     * Convert the uploads to UploadedFile, this will set the testing properties of UploadedFile
+     *
+     * @param Form $form
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: FQCN needed

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 3]
```diff
@@ -712,4 +730,27 @@ public function dump()
...
+     * @param array $uploads
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: Fix spacing.

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 4]
```diff
@@ -712,4 +730,27 @@ public function dump()
...
+        $names = array_keys($files);
+        $files = array_map(function (array $file, $name) use ($uploads) {
+            if (isset($uploads[$name])) {
+                $absolutePath = $uploads[$name];
+                $file = new UploadedFile($file['tmp_name'], basename($absolutePath), $file['type'], $file['size'], $file['error'], true);
+            }
+            return $file;
+        }, $files, $names);
+        $files = array_combine($names, $files);
+        return $files;
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: cs

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 5]
```diff
@@ -712,4 +730,27 @@ public function dump()
...
+
+        $names = array_keys($files);
+        $files = array_map(function (array $file, $name) use ($uploads) {
+            if (isset($uploads[$name])) {
+                $absolutePath = $uploads[$name];
+                $file = new UploadedFile($file['tmp_name'], basename($absolutePath), $file['type'], $file['size'], $file['error'], true);
+            }
+            return $file;
+        }, $files, $names);
+        $files = array_combine($names, $files);
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: cs

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 6]
```diff
@@ -712,4 +730,27 @@ public function dump()
...
+    {
+        $files = $form->getFiles();
+
+        $names = array_keys($files);
+        $files = array_map(function (array $file, $name) use ($uploads) {
+            if (isset($uploads[$name])) {
+                $absolutePath = $uploads[$name];
+                $file = new UploadedFile($file['tmp_name'], basename($absolutePath), $file['type'], $file['size'], $file['error'], true);
+            }
+            return $file;
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: cs

**Comment by jildertmiedema** [comment]
Some extra information about this PR:
I've created this solution to test a file upload.

The validation of the request failed on ```'picture' => 'required|mimes:jpeg,jpg',```, Symfony has some testing facilities in ```UploadedFile```, but they are not enabled by default.

The given solution parses the files in the request and converts them to ```UploadedFile``` instances, the validation will now pass because of the testing property of ```UploadedFile```.

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 7]
```diff
@@ -712,4 +728,31 @@ public function dump()
 
         dd($content);
     }
+
+    /**
+     * Convert the uploads to UploadedFile, this will set the testing properties of UploadedFile
+     *
+     * @param  Symfony\Component\DomCrawler\Form $form
+     * @param  array $uploads
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: cs

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 8]
```diff
@@ -712,4 +728,31 @@ public function dump()
 
         dd($content);
     }
+
+    /**
+     * Convert the uploads to UploadedFile, this will set the testing properties of UploadedFile
+     *
+     * @param  Symfony\Component\DomCrawler\Form $form
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: cs

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php (hunk 7 above)
Comment (jildertmiedema) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: Sorry, what am i missing?

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: 2 spaces before the $

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 9]
```diff
@@ -712,4 +728,31 @@ public function dump()
 
         dd($content);
     }
+
+    /**
+     * Convert the uploads to UploadedFile, this will set the testing properties of UploadedFile
+     *
+     * @param  \Symfony\Component\DomCrawler\Form $form
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: cs still wrong

File: src/Illuminate/Foundation/Testing/CrawlerTrait.php [hunk 10]
```diff
@@ -712,4 +728,31 @@ public function dump()
...
+    protected function convertUploads(Form $form, array $uploads)
+    {
+        $files = $form->getFiles();
+
+        $names = array_keys($files);
+
+        $files = array_map(function (array $file, $name) use ($uploads) {
+            if (isset($uploads[$name])) {
+                $absolutePath = $uploads[$name];
+                $file = new UploadedFile($file['tmp_name'], basename($absolutePath), $file['type'], $file['size'], $file['error'], true);
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/CrawlerTrait.php]: just return here

**Comment by ibrasho** [comment]
I think this should be squashed to one commit.

**Pull request merged**